File: Source/Workload/GtkSharp.Workload.Template.FSharp/content/GtkSharp.Application.FSharp/MainWindow.fs
namespace GtkNamespace

open Gtk

type MainWindow (builder : Builder) as this =
    inherit Window(builder.GetRawOwnedObject("MainWindow"))

    let mutable _label1 : Label = null
    let mutable _button1 : Button = null
    let mutable _counter = 0;

    do
        _label1 <- builder.GetObject("_label1") :?> Label
        _button1 <- builder.GetObject("_button1") :?> Button

        // Gtk 4 removed delete-event; CloseRequest is its replacement.
        this.CloseRequest.Add(fun args ->
            Application.Quit()
            args.RetVal <- false
        )
        _button1.Clicked.Add(fun _ ->
            _counter <- _counter + 1
            _label1.Text <- "Hello World! This button has been clicked " + _counter.ToString() + " time(s)."
        )

    new() = new MainWindow(new Builder("MainWindow.ui"))

File: Source/Templates/GtkSharp.Template.CSharp/content/GtkSharp.Application.CSharp/MainWindow.cs
using System;
using Gtk;
using UI = Gtk.Builder.ObjectAttribute;

namespace GtkNamespace
{
    class MainWindow : Window
    {
        [UI] private Label _label1 = null;
        [UI] private Button _button1 = null;

        private int _counter;

        public MainWindow() : this(new Builder("MainWindow.ui")) { }

        private MainWindow(Builder builder) : base(builder.GetRawOwnedObject("MainWindow"))
        {
            builder.Autoconnect(this);

            // Gtk 4 removed GtkWidget::delete-event. CloseRequest is the
            // signal a window gets when the user asks to close it; returning
            // false lets the default handler go ahead with the close.
            CloseRequest += Window_CloseRequest;
            _button1.Clicked += Button1_Clicked;
        }

        private void Window_CloseRequest(object sender, CloseRequestArgs a)
        {
            Application.Quit();
            a.RetVal = false;
        }

        private void Button1_Clicked(object sender, EventArgs a)
        {
            _counter++;
            _label1.Text = "Hello World! This button has been clicked " + _counter + " time(s).";
        }
    }
}
